Slow tests

IKSIN opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!-- 
    Don't forget about CHANGELOG! 
    
    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...
    
    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

<!-- Enumerate changes you made -->

## Verification

<!-- How you tested it? How do you know it works? -->

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -3,6 +3,7 @@ package store
 import (
 	"context"
 	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
```
fix imports pls

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -493,7 +495,158 @@ func TestProxyStore_SeriesSlowStores(t *testing.T) {
 		},
 		{
-			title: "partial response enabled one thanos query is slow to respond",
+			title: "partial response disabled 2/2 thanos query is slow to respond",
```
It's better to say: 
1st store is fast, 2nd is slow

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -493,7 +495,158 @@ func TestProxyStore_SeriesSlowStores(t *testing.T) {
+			expectedErr: errors.New("test: failed to receive any data in 4s from test: context deadline exceeded"),
+		},
+		{
+			title: "partial response disabled 2nd response of 1/2 thanos query is slow to respond",
```
add commas: 
partial response disabled, 1st store is slow, 2nd - fast

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -493,7 +495,158 @@ func TestProxyStore_SeriesSlowStores(t *testing.T) {
+			expectedErr: errors.New("test: failed to receive any data in 4s from test: context deadline exceeded"),
+		},
+		{
+			title: "partial response disabled 2nd response of 2/2 thanos query is slow to respond",
```
there is no difference with first test case (in title). Let's add something or remove this test?

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -554,6 +825,7 @@ func TestProxyStore_SeriesSlowStores(t *testing.T) {
 
 			seriesEquals(t, tc.expectedSeries, s.SeriesSet)
 			testutil.Equals(t, tc.expectedWarningsLen, len(s.Warnings), "got %v", s.Warnings)
+			testutil.Assert(t, t1.Sub(t0) < 5 * time.Second, "responseTimeout not working")
```
let's add more understandable message, smth like:
Series request has taken %d, expected: %d, it seems that responseTimeout doesn't work properly.

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -980,6 +1256,7 @@ type mockedStoreAPI struct {
 	RespLabelNames  *storepb.LabelNamesResponse
 	RespError       error
 	RespDuration    time.Duration
+	SlowRespIndex	int
```
var name tells nothing, I can't understand what will happen if I change it. 
Let's rename it somehow or add comment.
And do we really need it? Because we already passing stores as we want (e.g. we know which that 2nd store is slow)

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -1016,10 +1293,19 @@ type StoreSeriesClient struct {
 	i       int
 	respSet []*storepb.SeriesResponse
 	respDur time.Duration
+	slowRespIndex int
```
let's add comment here too

d-ulyanov commented on pkg/store/proxy_test.go:
```diff
@@ -1016,10 +1293,19 @@ type StoreSeriesClient struct {
 
 func (c *StoreSeriesClient) Recv() (*storepb.SeriesResponse, error) {
-	time.Sleep(c.respDur)
+	if c.respDur != 0 {
```
let's rewrite it a bit:
```
if c.respDur != 0 {
    switch {
    // sleep on each Recv() call
    case c.slowRespIndex == 0:
        time.Sleep()
    // Sleep on N Recv() call 
    case c.slowRespIndex != 0 && c.i == c.slowRespIndex
        time.Sleep()
    }
}

```

d-ulyanov replied:
or as option - you can extract func here:
```
if c.respDur != 0 {
    sleepBeforeRespond(c.slowRespIndex, c.i)
}
```

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -319,6 +378,11 @@ type streamSeriesSet struct {
 	closeSeries     context.CancelFunc
 }
 
+type RecvResp struct{
```
Add comments pls and make it private

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -329,6 +393,8 @@ func startStreamSeriesSet(
 	name string,
 	partialResponse bool,
 	responseTimeout time.Duration,
+	metrics *proxyStoreMetrics,
```
it would be better to accept prepared metric with labels

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -344,11 +410,49 @@ func startStreamSeriesSet(
+				t0 := time.Now()
+				r, err := s.stream.Recv() //long and block
+				t1 := time.Now()
+				if !metricsSended {
```
sent, not sended

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -344,11 +410,49 @@ func startStreamSeriesSet(
-
 		for {
-			r, err := s.stream.Recv()
+			cctx := ctx
```
let's rename this `ctx` to `singleReceiveCtx` or just `receiveCtx`

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -344,11 +410,49 @@ func startStreamSeriesSet(
+			var rr *RecvResp
+			go func() {
+				t0 := time.Now()
+				r, err := s.stream.Recv() //long and block
```
`// long and blocking`

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -344,11 +410,49 @@ func startStreamSeriesSet(
+				r, err := s.stream.Recv() //long and block
+				t1 := time.Now()
+				if !metricsSended {
+					metrics.firstRecvDuration.WithLabelValues(storeAddr).Observe(float64(t1.Sub(t0).Microseconds()))
```
let's use Seconds here

d-ulyanov replied:
https://prometheus.io/docs/practices/naming/

mkabischev replied:
`.Observe(time.Since(t0).Seconds())`

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -371,52 +475,30 @@ func startStreamSeriesSet(
 				continue
 			}
-
+			metrics.recvChannelSize.WithLabelValues(storeAddr).Observe(float64(len(s.recvCh)))
```
it's better to use prepared metrics with preset labels here too. Also - it will be much faster

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -371,52 +475,30 @@ func startStreamSeriesSet(
-
+			metrics.recvChannelSize.WithLabelValues(storeAddr).Observe(float64(len(s.recvCh)))
+			if len(s.recvCh) == 10 {
+				metrics.writesToBlockedChannel.WithLabelValues(storeAddr).Inc()
```
it seems that `channelBlockedDuration ` will be enough and I'm not sure we need this metric at all.
Also `== 10` is a magic number here :)

d-ulyanov replied:
let's remove it

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -371,52 +475,30 @@ func startStreamSeriesSet(
+			if len(s.recvCh) == 10 {
+				metrics.writesToBlockedChannel.WithLabelValues(storeAddr).Inc()
+			} else {
+				metrics.writesToFreeChannel.WithLabelValues(storeAddr).Inc()
```
and this one

d-ulyanov commented on pkg/store/proxy.go:
```diff
@@ -49,6 +58,50 @@ type ProxyStore struct {
+
+}
+
+func newProxyStoreMetrics(reg *prometheus.Registry) *proxyStoreMetrics {
```
`newProxyStoreMetrics` -> `registerProxyStoreMetrics`

d-ulyanov commented on cmd/thanos/query.go:
```diff
@@ -282,7 +282,7 @@ func runQuery(
 			unhealthyStoreTimeout,
 		)
-		proxy            = store.NewProxyStore(logger, stores.Get, component.Query, selectorLset, storeResponseTimeout)
+		proxy            = store.NewProxyStore(logger, stores.Get, component.Query, selectorLset, storeResponseTimeout, reg)
```
fix parameters sequence pls

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -256,10 +315,11 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 
+			seriesSet = append(seriesSet, startStreamSeriesSet(seriesCtx, s.logger, closeSeries,
+				wg, sc, respSender, st.String(), !r.PartialResponseDisabled, s.responseTimeout, s.metrics, st.Addr()))
+			st.String()
```
I think it can be removed

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +411,49 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		metricsSended := false
```
```
var metricsSent bool
```

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +411,49 @@ func startStreamSeriesSet(
-			r, err := s.stream.Recv()
+			cctx := ctx
+			var cancel context.CancelFunc
+			timeoutMsg := fmt.Sprintf("failed to receive any data from %s", s.name)
```
I don't like the idea to prepare error message in the very beginning. The only difference is in line #422: responseTimeout is added. But do we really need it? As I know it's configuration value. So it can be built on #442.

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -371,52 +476,30 @@ func startStreamSeriesSet(
 			select {
 			case s.recvCh <- r.GetSeries():
+				t1 := time.Now()
+				metrics.channelBlockedDuration.WithLabelValues(storeAddr).Observe(float64(t1.Sub(t0).Microseconds()))
```
`.Observe(time.Since(t0).Seconds())`

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +411,49 @@ func startStreamSeriesSet(
-
 		for {
-			r, err := s.stream.Recv()
+			cctx := ctx
```
I think we can reuse `ctx` variable here.

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +411,49 @@ func startStreamSeriesSet(
 		for {
-			r, err := s.stream.Recv()
+			cctx := ctx
+			var cancel context.CancelFunc
```
Let's move it to `if s.responseTimeout != 0 {` block. Next to its usage

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -10,6 +10,8 @@ import (
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
+
```
redundant line

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +386,53 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		var isFirstRecv bool = true
```
```go
isFirstRecv := true
```

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -319,6 +364,11 @@ type streamSeriesSet struct {
 	closeSeries     context.CancelFunc
 }
 
+type RecvResp struct {
```
Let's add comment for this struct because it's exported.

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -41,6 +42,17 @@ type Client interface {
 }
 
+type firstRecvMetrics struct {
+	payload   prometheus.Observer
```
Maybe smth like `withPayloadDuration` `withoutPayloadDuration` & `timeoutCount`?

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -49,12 +61,37 @@ type ProxyStore struct {
+		Help: "Timeout recv count.",
+	}, []string{"store"})
+
+	if reg != nil {
```
Minor: what about create `Register`/`RegisterIn` method in `* proxyStoreMetrics ` struct?
```go
func(m *proxyStoreMetrics) RegisterIn(r *prometheus.Registry) {
    r.MustRegister(m.firstRecvDuration)
    r.MustRegister(m.timeoutRecvCount)
}
```

And usage will look like:
```go
m := newProxyStoreMetrics()
m.RegisterIn(reg)
```

So there will be no optional argument in func

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -255,11 +295,16 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 				continue
 			}
+			metrics := &firstRecvMetrics{
+				noPayload: s.metrics.firstRecvDuration.WithLabelValues(st.Addr(), "0"),
```
minor: extract `0` & `1` to constants?

mkabischev commented on pkg/store/proxy_test.go:
```diff
@@ -1012,14 +1291,27 @@ func (s *mockedStoreAPI) LabelValues(ctx context.Context, req *storepb.LabelValu
+}
+
+func (c *StoreSeriesClient) sleepBeforeRespond() {
+	if c.slowSeriesIndex != 0 {
```
looks like can be replaced with:
```go
func (c *StoreSeriesClient) sleepBeforeRespond() {
    if c.slowSeriesIndex == 0 || c.slowSeriesIndex == c.i {
        time.Sleep(c.respDur)
    }
}
```

mkabischev replied:
But really I don't understand the logic here 🤣

mkabischev commented on pkg/store/proxy.go:
```diff
@@ -344,11 +395,54 @@ func startStreamSeriesSet(
+				if s.responseTimeout != 0 {
+					timeoutMsg = fmt.Sprintf("failed to receive any data in %s from %s", s.responseTimeout.String(), s.name)
+				}
+				err := errors.Wrap(ctx.Err(), timeoutMsg)
```
Maybe you can use:
```go
var err error

if s.responseTimeout == 0 {
    err = errors.Wrapf(ctx.Err(), ....)
} else {
    err = errors.Wrapf(ctx.Err(), ....)
}
```

simonpasquier committed:
ruler: fix the /api/v1/rules endpoint (#1773)

* test/e2e: test alerts and rules API endpoints

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

* pkg/rule: fix /api/v1/rules endpoint

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

* Address Bartek's comments

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

bwplotka committed:
Moved 0.9.0 release to 26th of Nov (#1782)

Due to KubeCon we need to delay the release to the next week ): It's already end of the week for most of us.

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

mdlayher committed:
pkg: fix staticcheck issues (#1784)

Signed-off-by: Matt Layher <mdlayher@gmail.com>

dependabot-preview[bot] committed:
build(deps): bump cloud.google.com/go from 0.48.0 to 0.49.0 (#1764)

Bumps [cloud.google.com/go](https://github.com/GoogleCloudPlatform/gcloud-golang) from 0.48.0 to 0.49.0.
- [Release notes](https://github.com/GoogleCloudPlatform/gcloud-golang/releases)
- [Changelog](https://github.com/googleapis/google-cloud-go/blob/master/CHANGES.md)
- [Commits](https://github.com/GoogleCloudPlatform/gcloud-golang/compare/v0.48.0...v0.49.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

simonpasquier committed:
test/e2e: start Alertmanager without clustering (#1786)

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
pkg/rule: retain original path for rule files (#1785)

This change ensures that the /api/v1/rules endpoint returns the original
path of the rule file instead of the path of the temporary file
generated by Thanos ruler.

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

kakkoyun committed:
Simplify prober (#1694)

* Simplify prober

Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

* Preserve existing behaviour

Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

* Add CHANGELOG entry

Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

simonpasquier committed:
pkg/rule: support identical rule filenames in different directories (#1791)

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

bwplotka committed:
Update prometheus for subqueries support in ruler. (#1790)

* Updated Prometheus dep to 2.14.0 tip (80ba03c67da11403fefa2d48c4dd30af0fee9fc0) for subqueries support in ruler.

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

* Added rule test against regression.

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

bwplotka committed:
Cut release v0.9.0-rc.0 (#1793)

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

daixiang0 committed:
delete broken link (#1799)

Signed-off-by: Xiang Dai <764524258@qq.com>

daixiang0 committed:
add --web.enable-admin-api comment (#1800)

* add --web.enable-admin-api comment

Signed-off-by: Xiang Dai <764524258@qq.com>

* remove white notes

Signed-off-by: Xiang Dai <764524258@qq.com>

Commit:
ProxyStore timeouts.

Signed-off-by: Aleksey Sin <asin@ozon.ru>
